### ISSUE TITLE
Move Mining Score to focus on censorship-detection rather than policy-alignment

### ISSUE THREAD

**Issue by TheBlueMatt** [mempool/mempool#4038]
In #3867, some of the discussion ended up being focused around whether the block audit score is best served with or without mempoolfullrbf. IMO that ended up being a bit of a distraction, and I admit forming the discussion around that point was somewhat my fault. Still, in the end I noted a completely alternative approach to the mining score that makes it substantially more able to detect miner misbehavior, without detecting other attacks or strange behavior which the miner is not related to (spuriously blaming the miner), and without forcing policy viewpoints onto miners:

> i think an alternative statement is if a block is "good" is "did all the transactions I had in the top 3MW of my mempool either get mined or conflicted". That way it doesn't matter which bitcoind is being run, all we care about is that there isn't something hanging around in the top of the mempool that for some (possibly very bad!) reason isn't being mined.

I'd like to suggest making that change here in a separate discussion, as its actually largely unrelated to #3867, and deserves to live on its own.

**Comment by mononaut** [comment]
This is more or less how the block audit feature works already, modulo imperfections in our heuristics and implementation.

Miners can choose to include whichever transactions they like, for whatever reason, without health score penalties.

But when miners do appear to be prioritizing by fee rate, and some higher rate transactions are conspicuously missing without a plausible explanation, then their score will reflect those missing transactions.

Our heuristic for this works by counting up the displaced weight X of any "added" transactions, and allowing the fee bands corresponding to the lowest X weight units of our projected template to be excluded without penalty (marked as "marginal fee rate", rather than "removed").

A good practical example is this recent block mined by ViaBTC:

https://mempool.space/block/00000000000000000002c5ac9d3575f253477e6268cc23a3cd0c49a68eea2e09

<img width="1093" alt="Screenshot 2023-07-24 at 11 01 26 AM" src="https://github.com/mempool/mempool/assets/83316221/cd7d5c8b-217c-4817-9cbd-78340d6c4015">

You can see that they've selected ~600kWU of their own low fee transactions without penalty (highlighted in blue), and a corresponding missing ~600kWU of transactions at the top of the projected block are marked "marginal fee rate" (indicated by the desaturated color).

The only negative factors in the health score for this block are the handful of small transactions highlighted in bright pink, which *are* missing without a good explanation. (a cursory investigation suggests they suffered from propagation issues, since they *still* haven't been mined, and can't be found on number of other explorers).

**Comment by mononaut** [comment]
> The only negative factors in the health score for this block are the handful of small transactions highlighted in bright pink, which are missing without a good explanation. (a cursory investigation suggests they suffered from propagation issues, since they still haven't been mined, and can't be found on number of other explorers

I suppose this is precisely an example of "spuriously blaming the miner" for something that isn't their fault, but it's also impossible to distinguish this case from a miner merely *pretending* not to have seen a transaction which has been consistently in the top 1-2MWU of our mempool for a while.

**Comment by mononaut** [comment]
1) We generally **do** penalize miners for excluding the expected version of a set of conflicting transactions, unless we observed that one of the conflicts did not signal BIP125 replaceability - this means miners should never be penalized for their choice of `mempoolfullrbf` alone.
  Conflicting transactions are also not penalized if our expected version was received within the last 3 minutes (recently broadcast), or fits one of our other heuristics like "marginal fee rate" etc.

2) Enabling `mempoolfullrbf=1` was more about maximizing the number of transactions our nodes see than improving mining scores (although there are some scenarios where full-rbf miners could have been inadvertently penalized due to indirect conflicts, whereas with `mempoolfullrbf=1` this should never happen).

3) Agreed. I think we are getting fairly close to 100% scores most of the time now, and the few transactions that *are* still marked as "removed" generally aren't obvious false positives.

Here's a nice example of the "recently broadcast" exception producing a 100% health score despite the block missing several hundred transactions received by our nodes in the preceding 2-3 minutes (highlighted in dull pink):

<img width="1097" alt="Screenshot 2023-07-24 at 2 04 14 PM" src="https://github.com/mempool/mempool/assets/83316221/b2df6f59-8f1d-401f-9130-dcdf12d8e6c7">

**Comment by mononaut** [comment]
yes, good point! in that case we should probably just allow all replacements.

**Comment by benpbolton** [comment]
This is... embarrassingly not the proper place for this feature request, but it would help me to quickly "grok" the incentives and honesty of the miner if the audit presented an 'expected block' fee total vs. 'actual block' fee total. 

Was the miner simply following the incentive structure (maximize fees)? How much did it vary from my node's ideal?

**Comment by benpbolton** [comment]
Indeed; I understand it's unreasonable to assume that *my* node's mempool is the same as the miner's (that is, arguably, the entire purpose of this audit tool). I think what I am curious about is whether *my* mempool's expected block would outperform the mined block in feerate. 

With a simple fee sum of my expected block, I have a quick intuitive (if not entirely accurate) measuring stick.

**It doesn't say:**

> The miner didn't make as much money due to:
> - bad luck or incompetence managing their mempool
> - shady back-room deals
> - censorship

But it *does* say the miner didn't make as much money as the block my node expected them to mine. Or... of more value to me still, if it shows the miner consistently making more in fees than I expect, then I might have a node connectivity or association or ruleset issue that I would want to address.

**Comment by benpbolton** [comment]
Totally; when we exit the space of "what happened" to "why did this happen" there are conjectures, assumptions, and (invariably) inaccuracies. To me: fee deviation is the true measure of mined block 'health'. All else is conjecture. 

I'll admit that it's quite appealing to have some UX guide me through "why is this different", but I suppose what I'm most interested in is "how is this different" from a sum-of-expected-fees-from-transactions perspective. We have the information already. And transactions fall into one of three categories:

1) they're included as expected
2) something was worth *more* and was included (my node didn't have updated info)
3) something was worth *less* and was included (their node didn't have updated info, they wanted to win a block race, backroom deal, censorship, etc)

all of the conjecture comes from assuming cause for #2, and #3 .. 

heh... annnnnnd I'm a dummy. The functionality is already built into `master` and I just didn't catch it

![Screenshot 2023-07-25 at 2 24 00 PM](https://github.com/mempool/mempool/assets/7146063/ff289dea-64a4-4818-a555-8e85a7fb9b61)

**Comment by benpbolton** [comment]
> there are plenty of reasons why a miner's mempool may not generate blocks with as large a fee as some other arbitrary mempool on the network...

Indeed. I think our primary disagreement is that you consider block health to be an (arbitrarily) score-able value based on tweaking things like "such and such a transaction was only seen within the last __ minutes, so it's reasonable if they haven't seen it yet", and I consider block health to be an objectively score-able value based on fee difference ("silly miner... they could've made 1% more in fees").

Some of that difference is that you're bleeding in _intent_ into whether a mined block is 'worse' than my mempool's expected block; and while your approach is quite reasonable (to bake in some rational interpretation of the deviation), it leaves quite a bit of editorial interpretation to the word 'health'.

The good news (for me) is that I apparently already have what I'm looking for in the total fee deviation. And the good news (for you) is that you will be able to continue to pursue a block health rating system based on some tweak-able values of whether it's reasonable that such-and-such a miner excludes such-and-such a transaction... and I will get to benefit from your interpreted values while still holding them against the objective difference of whether my node could've done better.

**Comment by mononaut** [comment]
returning to the original issue, after PR #4046 most mining pools are now scoring 100% more often than not:

<img width="572" alt="Screenshot 2023-07-27 at 9 26 32 AM" src="https://github.com/mempool/mempool/assets/83316221/7b16eae4-c6cf-4995-b514-1c5b8cf7f06a">

**Comment by mononaut** [comment]
> Nice! I think we should dig more into the transactions in question, then. Probably there's more cases we're missing if we're close but still missing a handful.

I've started categorizing [reasons for transactions to be missing from blocks](https://gist.github.com/mononaut/6cf51684bb4f4f488587f089ae08f92b).

The remaining false positives are now generally due to:
 - our high mempool limits (`maxmempool` and `mempoolexpiry`) keeping transactions that miners evict.
 - transaction relay issues, including stuff that would be fixed by package relay, as well as other propagation failures that may warrant further investigation.

**Comment by mononaut** [comment]
> I guess we just have to hard-code that logic and flag transactions as "likely evicted"? Can we lean on the second bitcoind that uses default maxmempool settings to calculate this?

The simplest cases are easy - check if missing transactions (or their ancestors) pay less than the current purge rate as determined by the second bitcoind, or are more than 2 weeks old.

The trickier part is detecting transactions that would have been evicted at some point in the past, and might not have been rebroadcast yet.

I suppose we can flag transactions at the point they would normally be evicted from default mempools, propagate that status to new descendants, and check for the flag during audits.

---

> Right, I think this is what's solved by accepting conflicts? Or are you saying things that might have been replace-able/replaced in other mempools, where the miner (presumably) has only an older much-lower-fee copy of a transaction?

Accepting conflicts solves a lot of these (I described them as "propagation splits" in the Gist).

I think there are some more convoluted scenarios where some part of a transaction's dependency graph exists in some mempools but not others (e.g. where conflicting versions of an ancestor or fee-bumping child exist).

The package relay scenario is when a transaction CPFP's an ancestor that pays below the current mempool minfee - if we had widespread package relay then miners with default mempools might be able to accept those transactions.

**Comment by mononaut** [comment]
We were also seeing a few transactions that just don't seem to have propagated well, despite paying reasonable fees and not obviously conflicting with anything else.

Our current theory is that it's related to how the 25 in-mempool ancestor/descendant limit interacts with different arrival orders of sibling transactions.

**Comment by mononaut** [comment]
currently about ~40% of transactions in the mempool have unconfirmed ancestors.

4-5% are involved in CPFP relationships (either as the child or a fee-bumped ancestor), rising to 10-15% in the top 4WU of the mempool.
